repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -186,7 +213,113 @@ private static String createQualifiedName(String parentDomainQualifiedName) thro
+            entityStore.createOrUpdate(entityStream, false); // adding new policy
+        } finally {
+            RequestContext.get().setSkipAuthorizationCheck(false);
+        }
+    }
 
+    private void updateDaapVisibilityPolicy(AtlasEntity newEntity, AtlasEntity currentEntity,  String currentProductDaapVisibility, String newProductDaapVisibility) throws AtlasBaseException{
+
+        AtlasEntity policy = new AtlasEntity();
+        policy.setTypeName(POLICY_ENTITY_TYPE);
+        policy.setAttribute(NAME,newEntity.getAttribute(NAME));
+        policy.setAttribute(QUALIFIED_NAME,currentEntity.getGuid()+"/read-policy");
+
+        switch (currentProductDaapVisibility) {
+            case PRIVATE:
+                switch (newProductDaapVisibility) {
+                    case PROTECTED:
+                        // create policy for policyUsers and policyGroups
+                        policy.setAttribute(ATTR_POLICY_USERS, newEntity.getAttribute(DAAP_VISIBILITY_USERS));
+                        policy.setAttribute(ATTR_POLICY_GROUPS, newEntity.getAttribute(DAAP_VISIBILITY_GROUPS));
+                        break;
+                    case PUBLIC:
+                        // set empty user list
+                        policy.setAttribute(ATTR_POLICY_USERS, Arrays.asList());
+                        // set user groups to public to allow access for all
+                        policy.setAttribute(ATTR_POLICY_GROUPS, Arrays.asList("public"));
+                }
+                break;
+            case PROTECTED:
+                switch (newProductDaapVisibility) {
+                    case PRIVATE:
+                        // create policy for policyUsers and policyGroups
+                        policy.setAttribute(ATTR_POLICY_USERS,Arrays.asList());
+                        policy.setAttribute(ATTR_POLICY_GROUPS,Arrays.asList());
+                        break;
+                    case PUBLIC:
+                        // set empty user list
+                        policy.setAttribute(ATTR_POLICY_USERS, Arrays.asList());
+                        // set user groups to public to allow access for all
+                        policy.setAttribute(ATTR_POLICY_GROUPS, Arrays.asList("public"));
+                }
+                break;
+            case PUBLIC:
+                switch (newProductDaapVisibility) {
+                    case PRIVATE:
+                        // create policy for policyUsers and policyGroups
+                        policy.setAttribute(ATTR_POLICY_USERS,Arrays.asList());
+                        policy.setAttribute(ATTR_POLICY_GROUPS,Arrays.asList());
+                        break;
+                    case PROTECTED:
+                        // create policy for policyUsers and policyGroups
+                        policy.setAttribute(ATTR_POLICY_USERS, newEntity.getAttribute(DAAP_VISIBILITY_USERS));
+                        policy.setAttribute(ATTR_POLICY_GROUPS, newEntity.getAttribute(DAAP_VISIBILITY_GROUPS));
```

Comment:
Let us create 3 helper methods for 3 cases to write these attributes, it will reduce code duplication 3 times